core/embed/rust/src/ui/model_tr/layout.rs

```diff
@@ -108,6 +159,146 @@ extern "C" fn new_confirm_text(n_args: usize, args: *const Obj, kwargs: *mut Map
     unsafe { util::try_with_args_and_kwargs(n_args, args, kwargs, block) }
 }
 
+extern "C" fn request_pin(n_args: usize, args: *const Obj, kwargs: *mut Map) -> Obj {
+    let block = |_args: &[Obj], kwargs: &Map| {
+        let prompt: StrBuffer = kwargs.get(Qstr::MP_QSTR_prompt)?.try_into()?;
+        let subprompt: StrBuffer = kwargs.get(Qstr::MP_QSTR_subprompt)?.try_into()?;
+        let allow_cancel: Option<bool> =
+            kwargs.get(Qstr::MP_QSTR_allow_cancel)?.try_into_option()?;
+        let shuffle: Option<bool> = kwargs.get(Qstr::MP_QSTR_shuffle)?.try_into_option()?;
+
+        let obj = LayoutObj::new(
+            PinPage::new(
+                prompt,
+                subprompt,
+                allow_cancel.unwrap_or(true),
+                shuffle.unwrap_or(false),
+            )
+            .into_child(),
+        )?;
+        Ok(obj.into())
+    };
+    unsafe { util::try_with_args_and_kwargs(n_args, args, kwargs, block) }
+}
+
+extern "C" fn show_share_words(n_args: usize, args: *const Obj, kwargs: *mut Map) -> Obj {
+    let block = |_args: &[Obj], kwargs: &Map| {
+        let share_words: StrBuffer = kwargs.get(Qstr::MP_QSTR_share_words)?.try_into()?;
+        let title = "Recovery seed";
+
+        let share_words_len = share_words.split(',').count();
+        let share_words_len_str: String<10> = String::from(share_words_len as i16);
+
+        let beginning_text: String<50> = util::format_string([
+            "Write down these ",
+            share_words_len_str.as_str(),
+            " words:\n\n",
+        ]);
+
+        let mut middle_words: String<500> = String::new();
+        for (index, word) in share_words.split(',').enumerate() {
+            let line: String<50> = util::format_string([
+                String::<2>::from(index as i16 + 1).as_str(),
+                ". ",
+                word,
+                "\n\n\n",
+            ]);
+
+            middle_words.push_str(&line).unwrap();
+        }
+
+        let end_text: String<50> = util::format_string([
+            "I wrote down all ",
+            share_words_len_str.as_str(),
+            " words in order.",
+        ]);
+
+        let text_to_show: String<600> = util::format_string([
```

Comment:
why add this together instead of installing separate paragraphs?

Reply:
I tried to create separate paragraphs, but then all the paragraphs need to have the same content-type - so all three would need to be `String<500>`, even though the beginning and end are very small.

I also did not succeed to combine the `Paragraphs` with `FormattedText`